src/Tree/Tree.js

```diff
@@ -1,238 +1,330 @@
+
+        const cellClassName = classnames(
+            'fd-tree__col',
+            className
+        );
+
+        return (
+            <div
+                {...rest}
+                className={cellClassName}>
+                {children}
+            </div>
+        );
+    }
+}
+
+TreeCell.propTypes = {
+    children: PropTypes.node,
+    className: PropTypes.string
+};
+
+export class TreeHeader extends Component {
+    render() {
+        const {
+            buttonProps,
+            children,
+            isExpanded,
+            onExpandAll,
+            ...rest
+        } = this.props;
+
+        return (
+            <div {...rest} className='fd-tree fd-tree--header'>
+                <div className='fd-tree__row fd-tree__row--header'>
+                    {
+                        React.Children.map(children, (child, index) => {
+                            const isFirstTreeCell = index === 0 && child.type && child.type.name === 'TreeCell';
+
+                            // Add control class to first TreeCell element
+                            const className = classnames({
+                                'fd-tree__col--control': isFirstTreeCell
+                            });
+
+                            // Add expand button to first TreeCell element
+                            const newChildren = isFirstTreeCell ? (
+                                <div>
+                                    <button
+                                        {...buttonProps}
+                                        aria-label='expand'
+                                        aria-pressed={isExpanded}
+                                        className='fd-tree__control'
+                                        onClick={onExpandAll} />
+                                    {child.props && child.props.children}
+                                </div>
+                            ) : child.props && child.props.children;
+
+                            return React.cloneElement(child, {
+                                children: newChildren,
+                                className
+                            });
+                        })
+                    }
+                </div>
+            </div>
+        );
+    }
+}
+
+TreeHeader.propTypes = {
+    buttonProps: PropTypes.object,
+    children: PropTypes.node,
+    isExpanded: PropTypes.bool,
+    onExpandAll: PropTypes.func
+};
+
+TreeHeader.propDescriptions = {
+    buttonProps: 'Additional props to be spread to the header expand/collapse `<button>` element.'
+};
+
+export class TreeRow extends Component {
     constructor(props) {
         super(props);
-        this.state = {
-            iStates: [],
-            expandAllClicked: false,
-            numberOfElements: 0
-        };
+
+        const {onExpandClick} = this.props;
+
+        // Generate unique id for row to manage expand/collapse state in parent
+        const id = shortid.generate();
+
+        this.state = {id};
```

Comment:
Storing an `id` in state seems wrong.  I understand not including it in the `render` function so it's not just regenerated each time, but how about:
```js
this.rowId = shortid.generate();
```
in the constructor?

Reply:
Fixed in fc864e3